tinygrad/codegen/linearizer.py

```diff
@@ -314,10 +314,7 @@ def linearize(self):
         fake_global_idxs = [x*0 for x in global_idxs]
         self.global_store(-1, fake_global_idxs+local_idxs+fake_reduce_idxs+upcast_idxs, acc)  # store accumulators
         self.uop(UOps.BARRIER, None, [], ())
-        self.uop(UOps.ENDLOOP, None, [], (loop_local_idxs, "local"))
-
-        # local indexs are over, 0 them out
-        local_idxs = [x*0 for x in local_idxs]
+        self.uop(UOps.ENDLOOP, None, [], ([loop_local_idxs[i]*0 if i < len(self.group_for_reduce) else loop_local_idxs[i] for i in range(0,self.local_dims+len(self.group_for_reduce))], "local"))
```

Comment:
just `self.uop(UOps.ENDLOOP, None, [], ([loop_local_idxs[i] for i in range(0,self.local_dims+len(self.group_for_reduce)) if i >= len(self.group_for_reduce)], "local"))` seems fixing asm backends.

Are you also sure we not need to zero out local indices which are grouped_reduce?

Reply:
Updated my old patch with similar functionality, it's rebased on top of your changes: https://github.com/nimlgen/tinygrad/commit/67c062cd01fad73e69908a3a8702d60eb7d88fd6. This one might help you, so feel free to use this if you find this useful.

Reply:
Yes this is necessary to have the correct `if (lidx2==0) {` for the late reduce.  Zero'ing them all produces `if (0==0)` which is technically okay but calculates and overwrites device memory `lidx2` times more than necessary.   The old logic with `loop_local_idxs` would produce `if ((lidx1+lidx2) == 0) {` which isn't correct since we need `lidx1` late reduces.  Hope that makes sense.

Reply:
Not used anymore.